{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <img src=\"../img/refgenie_logo.svg\" class=\"img-header\"> hello world! \n",
    "\n",
    "To test `refgenie`, we'll index the human mitochondrial genome. Just run this command:\n",
    "\n",
    "Without docker: `refgenie -i http://big.databio.org/example_data/rCRS.fa.gz`\n",
    "\n",
    "With docker: `refgenie -i http://big.databio.org/example_data/rCRS.fa.gz -d`\n",
    "\n",
    "But first, let's look at the `--help` output to get an idea of usage options:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: refgenie [-h] [-R] [-C CONFIG_FILE] -i INPUT [-n NAME] [-a ANNOTATION]\n",
      "                [-d] [-o OUTFOLDER]\n",
      "\n",
      "Pipeline\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  -R, --recover         Overwrite locks to recover from previous failed run\n",
      "  -C CONFIG_FILE, --config CONFIG_FILE\n",
      "                        Pipeline configuration file (YAML). Relative paths are\n",
      "                        with respect to the pipeline script.\n",
      "  -i INPUT, --input INPUT\n",
      "                        Local path or URL to genome sequence file in .fa,\n",
      "                        .fa.gz, or .2bit format.\n",
      "  -n NAME, --name NAME  Name of the genome to build. If ommitted, refgenie\n",
      "                        will usethe basename of the file specified in --input\n",
      "  -a ANNOTATION, --annotation ANNOTATION\n",
      "                        Path to GTF gene annotation file\n",
      "  -d, --docker          Run all commands in the refgenie docker container.\n",
      "  -o OUTFOLDER, --outfolder OUTFOLDER\n",
      "                        Path to output genomes folder, using the $GENOMES\n",
      "                        environment variableif set. Currently set to:\n",
      "                        '/ext/yeti/genomes/'\n"
     ]
    }
   ],
   "source": [
    "refgenie --help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's go ahead and explore the output. This will produce a lot of logging information from all the indexers. Scroll down past all this to get to the next section where we'll explore the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using genome name: rCRS\n",
      "('Output to: ', 'rCRS', '/ext/yeti/genomes/', '/ext/yeti/genomes/rCRS')\n",
      "/home/nsheff/.local/lib/python3.5/site-packages/refgenie/refgenie.yaml\n",
      "### [Pipeline run code and environment:]\n",
      "\n",
      "*              Command:  `/home/nsheff/.local/bin/refgenie -i http://big.databio.org/example_data/rCRS.fa.gz -d`\n",
      "*         Compute host:  puma\n",
      "*          Working dir:  /home/nsheff/code/refgenie/docs_jupyter\n",
      "*            Outfolder:  /ext/yeti/genomes/rCRS/\n",
      "*  Pipeline started at:   (05-10 13:48:13) elapsed: 0.0 _TIME_\n",
      "\n",
      "### [Version log:]\n",
      "\n",
      "*       Python version:  3.5.2\n",
      "*          Pypiper dir:  `/home/nsheff/.local/lib/python3.5/site-packages/pypiper`\n",
      "*      Pypiper version:  0.11dev\n",
      "*         Pipeline dir:  `/home/nsheff/.local/bin`\n",
      "*     Pipeline version:  None\n",
      "\n",
      "### [Arguments passed to pipeline:]\n",
      "\n",
      "*         `annotation`:  `None`\n",
      "*          `outfolder`:  `/ext/yeti/genomes/`\n",
      "*              `input`:  `http://big.databio.org/example_data/rCRS.fa.gz`\n",
      "*        `config_file`:  `/home/nsheff/.local/lib/python3.5/site-packages/refgenie/refgenie.yaml`\n",
      "*            `recover`:  `False`\n",
      "*             `docker`:  `True`\n",
      "*               `name`:  `None`\n",
      "\n",
      "----------------------------------------\n",
      "\n",
      "\n",
      "Changed status from initializing to running.\n",
      "\n",
      "Loading config file: /home/nsheff/.local/lib/python3.5/site-packages/refgenie/refgenie.yaml\n",
      "\n",
      "Target to produce: `/ext/yeti/genomes/rCRS/rCRS.fa.gz`\n",
      "\n",
      "\n",
      "> `wget -O /ext/yeti/genomes/rCRS/rCRS.fa.gz http://big.databio.org/example_data/rCRS.fa.gz` (23384)\n",
      "\n",
      "<pre>\n",
      "--2019-05-10 13:48:14--  http://big.databio.org/example_data/rCRS.fa.gz\n",
      "Resolving big.databio.org (big.databio.org)... 128.143.8.170\n",
      "Connecting to big.databio.org (big.databio.org)|128.143.8.170|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 5316 (5.2K) [application/octet-stream]\n",
      "Saving to: ‘/ext/yeti/genomes/rCRS/rCRS.fa.gz’\n",
      "\n",
      "     0K .....                                                 100%  152M=0s\n",
      "\n",
      "2019-05-10 13:48:14 (152 MB/s) - ‘/ext/yeti/genomes/rCRS/rCRS.fa.gz’ saved [5316/5316]\n",
      "\n",
      "</pre>\n",
      "Process returns and memory:23384: 0 (0.004GB);  Elapsed: 0:00:00. Peak memory: 0.004GB.\n",
      "\n",
      "> `docker run -itd -v /ext/yeti/genomes/rCRS:/ext/yeti/genomes/rCRS nsheff/refgenie`\n",
      "\n",
      "Using docker container: 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a\n",
      "Target to produce: `/ext/yeti/genomes/rCRS/rCRS.fa`\n",
      "\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a gzip -f -cd /ext/yeti/genomes/rCRS/rCRS.fa.gz > /ext/yeti/genomes/rCRS/rCRS.fa` (23482)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23482: 0 (0.017GB);  Elapsed: 0:00:00. Peak memory: 0.017GB.\n",
      "Target to produce: `/ext/yeti/genomes/rCRS/rCRS.fa.fai`\n",
      "\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a samtools faidx /ext/yeti/genomes/rCRS/rCRS.fa` (23518)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23518: 0 (0.017GB);  Elapsed: 0:00:00. Peak memory: 0.017GB.\n",
      "Target to produce: `/ext/yeti/genomes/rCRS/rCRS.chromSizes`\n",
      "\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a cut -f 1,2 /ext/yeti/genomes/rCRS/rCRS.fa.fai > /ext/yeti/genomes/rCRS/rCRS.chrom.sizes` (23552)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23552: 0 (0.018GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a ln -s rCRS.chrom.sizes /ext/yeti/genomes/rCRS/rCRS.chromSizes` (23587)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23587: 0 (0.017GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "* No GTF gene annotations provided. Skipping this step.\n",
      "Target to produce: `/ext/yeti/genomes/rCRS/indexed_bowtie2/completed.flag`\n",
      "\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a ln -sf ../rCRS.fa /ext/yeti/genomes/rCRS/indexed_bowtie2` (23622)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23622: 0 (0.018GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a bowtie2-build /ext/yeti/genomes/rCRS/rCRS.fa /ext/yeti/genomes/rCRS/indexed_bowtie2/rCRS` (23660)\n",
      "\n",
      "<pre>\n",
      "Settings:\n",
      "  Output files: \"/ext/yeti/genomes/rCRS/indexed_bowtie2/rCRS.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /ext/yeti/genomes/rCRS/rCRS.fa\n",
      "Reading reference sizes\n",
      "Building a SMALL index\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 4142\n",
      "Using parameters --bmax 3107 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 3107 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 16568 (target: 3106)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 16568 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 16569 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 5124\n",
      "fchr[G]: 10305\n",
      "fchr[T]: 12474\n",
      "fchr[$]: 16568\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4200050 bytes to primary EBWT file: /ext/yeti/genomes/rCRS/indexed_bowtie2/rCRS.1.bt2\n",
      "Wrote 4148 bytes to secondary EBWT file: /ext/yeti/genomes/rCRS/indexed_bowtie2/rCRS.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 16568\n",
      "    bwtLen: 16569\n",
      "    sz: 4142\n",
      "    bwtSz: 4143\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 1036\n",
      "    offsSz: 4144\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 87\n",
      "    numLines: 87\n",
      "    ebwtTotLen: 5568\n",
      "    ebwtTotSz: 5568\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 4142\n",
      "Using parameters --bmax 3107 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 3107 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 16568 (target: 3106)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 16568 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 16569 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 5124\n",
      "fchr[G]: 10305\n",
      "fchr[T]: 12474\n",
      "fchr[$]: 16568\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4200050 bytes to primary EBWT file: /ext/yeti/genomes/rCRS/indexed_bowtie2/rCRS.rev.1.bt2\n",
      "Wrote 4148 bytes to secondary EBWT file: /ext/yeti/genomes/rCRS/indexed_bowtie2/rCRS.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 16568\n",
      "    bwtLen: 16569\n",
      "    sz: 4142\n",
      "    bwtSz: 4143\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 1036\n",
      "    offsSz: 4144\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 87\n",
      "    numLines: 87\n",
      "    ebwtTotLen: 5568\n",
      "    ebwtTotSz: 5568\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:00\n",
      "</pre>\n",
      "Process returns and memory:23660: 0 (0.018GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a touch /ext/yeti/genomes/rCRS/indexed_bowtie2/completed.flag` (23698)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23698: 0 (0.017GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "Target to produce: `/ext/yeti/genomes/rCRS/indexed_hisat2/completed.flag`\n",
      "\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a ln -sf ../rCRS.fa /ext/yeti/genomes/rCRS/indexed_hisat2` (23730)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23730: 0 (0.018GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a hisat2-build /ext/yeti/genomes/rCRS/rCRS.fa /ext/yeti/genomes/rCRS/indexed_hisat2/rCRS` (23766)\n",
      "\n",
      "<pre>\n",
      "Settings:\n",
      "  Output files: \"/ext/yeti/genomes/rCRS/indexed_hisat2/rCRS.*.ht2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Local offset rate: 3 (one in 8)\n",
      "  Local fTable chars: 6\n",
      "  Local sequence length: 57344\n",
      "  Local sequence overlap between two consecutive indexes: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /ext/yeti/genomes/rCRS/rCRS.fa\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to read SNPs and splice sites: 00:00:00\n",
      "Using parameters --bmax 3107 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 3107 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 2070.12 (target: 3106)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering GFM loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (3107) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 2134 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2135 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (3107) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 2033 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2034 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (3107) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 2198 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2199 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (3107) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 1602 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1603 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (3107) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 1846 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1847 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (3107) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 2428 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2429 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (3107) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 2792 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 2793 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (3107) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 1528 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 1529 for bucket 8\n",
      "Exited GFM loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 5124\n",
      "fchr[G]: 10305\n",
      "fchr[T]: 12474\n",
      "fchr[$]: 16568\n",
      "Exiting GFM::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 4200070 bytes to primary GFM file: /ext/yeti/genomes/rCRS/indexed_hisat2/rCRS.1.ht2\n",
      "Wrote 4148 bytes to secondary GFM file: /ext/yeti/genomes/rCRS/indexed_hisat2/rCRS.2.ht2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from GFM constructor\n",
      "Returning from initFromVector\n",
      "Wrote 13035 bytes to primary GFM file: /ext/yeti/genomes/rCRS/indexed_hisat2/rCRS.5.ht2\n",
      "Wrote 4148 bytes to secondary GFM file: /ext/yeti/genomes/rCRS/indexed_hisat2/rCRS.6.ht2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Re-opening _in5 and _in5 as input streams\n",
      "Returning from HierEbwt constructor\n",
      "Headers:\n",
      "    len: 16568\n",
      "    gbwtLen: 16569\n",
      "    nodes: 16569\n",
      "    sz: 4142\n",
      "    gbwtSz: 4143\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 0\n",
      "    eftabSz: 0\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 1036\n",
      "    offsSz: 4144\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideGbwtSz: 48\n",
      "    sideGbwtLen: 192\n",
      "    numSides: 87\n",
      "    numLines: 87\n",
      "    gbwtTotLen: 5568\n",
      "    gbwtTotSz: 5568\n",
      "    reverse: 0\n",
      "    linearFM: Yes\n",
      "Total time for call to driver() for forward index: 00:00:00\n",
      "</pre>\n",
      "Process returns and memory:23766: 0 (0.017GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a touch /ext/yeti/genomes/rCRS/indexed_hisat2/completed.flag` (23803)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23803: 0 (0.017GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "Target to produce: `/ext/yeti/genomes/rCRS/indexed_kallisto/completed.flag`\n",
      "\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a kallisto index -i /ext/yeti/genomes/rCRS/indexed_kallisto/rCRS_kallisto_index.idx /ext/yeti/genomes/rCRS/rCRS.fa` (23838)\n",
      "\n",
      "<pre>\n",
      "\n",
      "[build] loading fasta file /ext/yeti/genomes/rCRS/rCRS.fa\n",
      "[build] k-mer length: 31\n",
      "[build] warning: replaced 1 non-ACGUT characters in the input sequence\n",
      "        with pseudorandom nucleotides\n",
      "[build] counting k-mers ... done.\n",
      "[build] building target de Bruijn graph ...  done \n",
      "[build] creating equivalence classes ...  done\n",
      "[build] target de Bruijn graph has 1 contigs and contains 16539 k-mers \n",
      "\n",
      "</pre>\n",
      "Process returns and memory:23838: 0 (0.018GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "\n",
      "> `docker exec 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a touch /ext/yeti/genomes/rCRS/indexed_kallisto/completed.flag` (23873)\n",
      "\n",
      "<pre>\n",
      "</pre>\n",
      "Process returns and memory:23873: 0 (0.018GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n",
      "\n",
      "Changed status from running to completed.\n",
      "Starting cleanup: 0 files; 0 conditional files for cleanup\n",
      "\n",
      "> `Time`\t0:00:03\trefgenie\t_RES_\n",
      "\n",
      "> `Success`\t05-10-13:48:16\trefgenie\t_RES_\n",
      "\n",
      "##### [Epilogue:]\n",
      "*   Total elapsed time:  0:00:03\n",
      "*     Peak memory used:  0.02 GB\n",
      "* Pipeline completed at:  (05-10 13:48:16) elapsed: 3.0 _TIME_\n",
      "Removing docker container. . .\n",
      "\n",
      "> `docker rm -f 80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a` (23906)\n",
      "\n",
      "<pre>\n",
      "80b490fa078b89317c18e0c0df8bcd40530a605c0dbdb65fee77ca7c8f3b984a\n",
      "</pre>\n",
      "Process returns and memory:23906: 0 (0.018GB);  Elapsed: 0:00:00. Peak memory: 0.018GB.\n"
     ]
    }
   ],
   "source": [
    "refgenie -i http://big.databio.org/example_data/rCRS.fa.gz -d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We were able to index that mitochondrial genome in just a matter of seconds (full genomes will obviously take much longer). Let's see what this process produced:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m/ext/yeti/genomes/rCRS\u001b[00m\n",
      "├── \u001b[01;34mindexed_bowtie2\u001b[00m\n",
      "│   ├── completed.flag\n",
      "│   ├── rCRS.1.bt2\n",
      "│   ├── rCRS.2.bt2\n",
      "│   ├── rCRS.3.bt2\n",
      "│   ├── rCRS.4.bt2\n",
      "│   ├── \u001b[01;36mrCRS.fa\u001b[00m -> ../rCRS.fa\n",
      "│   ├── rCRS.rev.1.bt2\n",
      "│   └── rCRS.rev.2.bt2\n",
      "├── \u001b[01;34mindexed_hisat2\u001b[00m\n",
      "│   ├── completed.flag\n",
      "│   ├── rCRS.1.ht2\n",
      "│   ├── rCRS.2.ht2\n",
      "│   ├── rCRS.3.ht2\n",
      "│   ├── rCRS.4.ht2\n",
      "│   ├── rCRS.5.ht2\n",
      "│   ├── rCRS.6.ht2\n",
      "│   ├── rCRS.7.ht2\n",
      "│   ├── rCRS.8.ht2\n",
      "│   └── \u001b[01;36mrCRS.fa\u001b[00m -> ../rCRS.fa\n",
      "├── \u001b[01;34mindexed_kallisto\u001b[00m\n",
      "│   ├── completed.flag\n",
      "│   └── rCRS_kallisto_index.idx\n",
      "├── rCRS.chrom.sizes\n",
      "├── \u001b[01;36mrCRS.chromSizes\u001b[00m -> rCRS.chrom.sizes\n",
      "├── rCRS.fa\n",
      "├── rCRS.fa.fai\n",
      "├── \u001b[01;31mrCRS.fa.gz\u001b[00m\n",
      "├── refgenie_commands.sh\n",
      "├── refgenie_completed.flag\n",
      "├── refgenie_log.md\n",
      "├── refgenie_profile.tsv\n",
      "└── stats.tsv\n",
      "\n",
      "3 directories, 30 files\n"
     ]
    }
   ],
   "source": [
    "tree /ext/yeti/genomes/rCRS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here you can see that we have indexes for a few common indexers. All of the above output will be saved in the `refgenie_log.md` file."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
